### PR TITLE
update existing github gist

### LINKED ISSUE #153: feature: mechanism for updating gists
If a user is the owner of a gist and they update it (maybe even a small tweak like adding comments), it would be nice to allow them to be able to update the gist (create a revision) rather than creating a new gist. In order to allow this option in the "Save Project" dialog we would somehow need to keep track of what the source gist was for the project. I don't think storing that in the meta.json of the project is the right thing to do -- maybe there's another way.

### PR DESCRIPTION
**Pull request by magland**
In save project window, user can either save to new gist or update existing gist. When updating the existing Gist, they need to manually enter the Gist URL (`https://gist.github.com/<user>/<id>`). For this PR we don't keep track of the origin gist URL for the open project (even though it would be convenient) because there are a number of questions about how that would be done. I think that should be a separate PR.

Closes #153

### CODE REVIEW

File: gui/src/app/gists/saveAsGitHubGist.ts [hunk 1]
```diff
@@ -67,6 +67,7 @@ export const updateGitHubGist = async (
       files[path] = { content };
     }
   }
+  console.log("--------------- files", files);
```
Comment (jsoules) [inline gui/src/app/gists/saveAsGitHubGist.ts:70]: Is this intended to be in the production version?

File: gui/src/app/pages/HomePage/SaveProjectWindow.tsx [hunk 2]
```diff
@@ -204,6 +283,52 @@ const GistExportView: FunctionComponent<GistExportViewProps> = ({
...
+  }
+  // handle deleted files
+  for (const fname in existingFiles) {
+    if (!newFiles[fname] || !newFiles[fname].trim()) {
+      patch[fname] = null;
+    }
+  }
+  return patch;
+};
+
```
Comment (jsoules) [inline gui/src/app/pages/HomePage/SaveProjectWindow.tsx:331]: Gist-manipulation logic should go in the `gists/SaveAsGithubGist` rather than in the UI component

**Review by jsoules — CHANGES_REQUESTED** [review]

File: gui/src/app/gists/saveAsGitHubGist.ts (hunk 1 above)
Comment (magland) [reply gui/src/app/gists/saveAsGitHubGist.ts]: fixed

File: gui/src/app/pages/HomePage/SaveProjectWindow.tsx (hunk 2 above)
Comment (magland) [reply gui/src/app/pages/HomePage/SaveProjectWindow.tsx]: Done

**Comment by magland** [comment]
This is going to have a conflict with #158, so after one of these is merged I'll work on resolving it.

**Comment by magland** [comment]
I'm going to need to do increasingly complex conflict resolution for this and #158 once the latest PRs are merged. Is there a reason why these two are not getting merged? I just would like to know when I should incorporate those changes.

@WardBrian @jsoules

**Comment by WardBrian** [comment]
I don't have any issue with the functionality of this PR, but I do find the logic to create the `GistExplainer` div to be a bit hard to follow now

I haven't looked at #158 in a few days. I'm not entirely convinced it's a good functionality to provide (it seems almost certain to me that people will misunderstand it and lose their data), but I'm not so opposed as to prevent it being included

**Comment by magland** [comment]
Okay I'll work on resolving conflicts for this one first, and try to clarify/simplify the GistExplainer

**Comment by magland** [comment]
I've resolved the conflicts here. @WardBrian could you elaborate a bit on where you are confused? Are you talking about this code?

https://github.com/flatironinstitute/stan-playground/blob/8476c9652453341228aa2b684d8843d51c07ec0f/gui/src/app/pages/HomePage/SaveProjectWindow.tsx#L51-L94

**Comment by WardBrian** [comment]
No, it's really more the code on line 169-on that actually builds up the view -- I understand the desire to share between the uploading and updating case, but it leads to a lot of conditionals and even conditional prose. 

I think it would make sense to split them, because a user who is updating an existing gist has already been through the token creation, presumably?

**Comment by magland** [comment]
> No, it's really more the code on line 169-on that actually builds up the view -- I understand the desire to share between the uploading and updating case, but it leads to a lot of conditionals and even conditional prose.
> 
> I think it would make sense to split them, because a user who is updating an existing gist has already been through the token creation, presumably?

Okay I'll take a crack at that.

**Review by jsoules — APPROVED** [review]
I'm okay moving forward with this.

**Comment by magland** [comment]
@WardBrian I made a separate component for updating the gist, and tried to reuse components where it seemed appropriate.

**Comment by WardBrian** [comment]
Thanks -- looks great to me!

**Review by WardBrian — APPROVED** [review]
